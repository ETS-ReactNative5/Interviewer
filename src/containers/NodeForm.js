import React, { Component } from 'react';
import { bindActionCreators, compose } from 'redux';
import { connect } from 'react-redux';
import { submit } from 'redux-form';
import { debounce } from 'lodash';
import { Button } from '@codaco/ui';
import Overlay from './Overlay';
import Form from './Form';
import FormWizard from './FormWizard';
import { entityAttributesProperty } from '../ducks/modules/network';
import { Scroller } from '../components';

const reduxFormName = 'NODE_FORM';

class NodeForm extends Component {
  constructor(props) {
    super(props);

    this.overlay = React.createRef();
  }

  handleSubmit = debounce((form) => {
    this.props.onSubmit({ form });
    this.props.onClose();
  }, 1000, { // This is needed to prevent double submit.
    leading: true,
    trailing: false,
  })

  render() {
    const { show, form, initialValues, submitForm, stage } = this.props;

    const formProps = {
      ...form,
      initialValues,
      onSubmit: this.handleSubmit,
      autoFocus: true,
      subject: stage.subject,
      form: reduxFormName,
    };

    return (
      <Overlay
        show={show}
        title={form.title}
        onClose={this.props.onClose}
        className="node-form"
      >
        { this.props.useFullScreenForms ?
          <FormWizard
            {...formProps}
          /> :
          <React.Fragment>
            <Scroller>
              <Form
                {...formProps}
              />
            </Scroller>
            <div className="node-form__footer">
              <Button key="submit" aria-label="Submit" type="submit" onClick={submitForm}>
                Finished
              </Button>
            </div>
          </React.Fragment>
        }
      </Overlay>
    );
  }
}

const mapStateToProps = (state, props) => {
  const nodeAttributes = props.node ? props.node[entityAttributesProperty] : {};

  const initialValues = {
    ...nodeAttributes,
  };

  return {
    form: props.stage.form,
    useFullScreenForms: state.deviceSettings.useFullScreenForms,
    initialValues,
  };
};

const mapDispatchToProps = dispatch => ({
  submitForm: bindActionCreators(() => submit(reduxFormName), dispatch),
});

export { NodeForm };

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
)(NodeForm);
