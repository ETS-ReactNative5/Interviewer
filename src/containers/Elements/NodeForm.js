import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { reset } from 'redux-form';
import PropTypes from 'prop-types';
import { pick, map } from 'lodash';
import { createSelector } from 'reselect';
import { actionCreators as modalActions } from '../../ducks/modules/modals';
import { Form } from '../../containers/Elements';
import { Modal } from '../../components/Elements';
import { makeRehydrateFields } from '../../selectors/rehydrate';

const propNode = (_, props) => props.node;

const makePropFieldVariables = () =>
  createSelector(
    makeRehydrateFields(),
    fields => map(fields, 'name'),
  );

const makeGetInitialValuesFromProps = () =>
  createSelector(
    makePropFieldVariables(),
    propNode,
    (fields, node) => pick(node, fields),
  );

/**
  * Modal Node Form, than can handle new/editing of nodes
  * @extends Component
  */
class NodeForm extends Component {
  onSubmit = (formData, dispatch, form) => {
    this.props.closeModal(this.props.name);
    this.props.handleSubmit(formData, dispatch, form);
    if (this.state.typeOfSubmit === 'continuous') {
      this.props.resetValues(form.name);
      this.props.openModal(this.props.name);
    }
  };

  continuousSubmit = () => {
    this.setState({
      typeOfSubmit: 'continuous',
    }, this.submit);
  };

  normalSubmit = () => {
    this.setState({
      typeOfSubmit: 'normal',
    }, this.submit);
  };

  render() {
    const {
      title,
      fields,
      name,
      addAnother,
      initialValues,
    } = this.props;

    return (
      <Modal name={name} title={title}>
        <Form
          fields={fields}
          initialValues={!addAnother ? initialValues : null}
          autoFocus
          form={name.toString()}
          onSubmit={this.onSubmit}
          addAnother={addAnother}
          continuousSubmit={this.continuousSubmit}
          normalSubmit={this.normalSubmit}
        />
      </Modal>
    );
  }
}

NodeForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  openModal: PropTypes.func.isRequired,
  resetValues: PropTypes.func.isRequired,
  initialValues: PropTypes.any.isRequired,
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.symbol,
  ]).isRequired,
  title: PropTypes.string.isRequired,
  fields: PropTypes.array.isRequired,
  node: PropTypes.any, // eslint-disable-line react/no-unused-prop-types
  addAnother: PropTypes.bool,
};

NodeForm.defaultProps = {
  addAnother: false,
  node: {},
};

function makeMapStateToProps() {
  const getInitialValuesFromProps = makeGetInitialValuesFromProps();

  return function mapStateToProps(state, props) {
    return {
      initialValues: getInitialValuesFromProps(state, props),
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    closeModal: bindActionCreators(modalActions.closeModal, dispatch),
    openModal: bindActionCreators(modalActions.openModal, dispatch),
    resetValues: bindActionCreators(reset, dispatch),
  };
}

export default connect(makeMapStateToProps, mapDispatchToProps)(NodeForm);
