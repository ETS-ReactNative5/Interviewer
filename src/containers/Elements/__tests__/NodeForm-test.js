/* eslint-env jest */

import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import NodeForm from '../NodeForm';

const node = {
  foo: 'bar',
};

const registry = {
  foo: {
    type: 'string',
    label: 'Foo',
  },
};

const mockStore = () =>
  createStore(
    () => (
      {
        protocol: {
          config: {
            registry,
          },
        },
        modals: [
          { name: 'baz', open: true },
        ],
      }
    ),
  );

const mockProps = {
  name: 'baz',
  title: 'The form title',
  handleSubmit: () => {},
  fields: [{
    variable: 'foo',
    component: 'TextInput',
  }],
};

describe('<NodeForm />', () => {
  it('should render', () => {
    const subject = shallow((
      <NodeForm {...mockProps} store={mockStore()} />
    ));

    expect(subject).toMatchSnapshot();
  });

  it('should render with prepopulated fields if provided', () => {
    const subject = mount((
      <Provider store={mockStore()}>
        <NodeForm {...mockProps} node={node} />
      </Provider>
    ));

    expect(subject.find('Form').first().prop('initialValues')).toEqual({ foo: 'bar' });
  });
});
