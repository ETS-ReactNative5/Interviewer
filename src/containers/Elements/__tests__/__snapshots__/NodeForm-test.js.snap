// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeForm /> should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <NodeForm
    addAnother={false}
    closeModal={[Function]}
    fields={
        Array [
            Object {
              "component": "TextInput",
              "variable": "foo",
            },
          ]
    }
    handleSubmit={[Function]}
    initialValues={Object {}}
    name="baz"
    node={Object {}}
    openModal={[Function]}
    resetValues={[Function]}
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
    title="The form title"
/>,
  "nodes": Array [
    <NodeForm
      addAnother={false}
      closeModal={[Function]}
      fields={
            Array [
                  Object {
                    "component": "TextInput",
                    "variable": "foo",
                  },
                ]
      }
      handleSubmit={[Function]}
      initialValues={Object {}}
      name="baz"
      node={Object {}}
      openModal={[Function]}
      resetValues={[Function]}
      store={
            Object {
                  "dispatch": [Function],
                  "getState": [Function],
                  "replaceReducer": [Function],
                  "subscribe": [Function],
                  Symbol(observable): [Function],
                }
      }
      title="The form title"
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Connect(NodeForm)
        fields={
                Array [
                        Object {
                          "component": "TextInput",
                          "variable": "foo",
                        },
                      ]
        }
        handleSubmit={[Function]}
        name="baz"
        store={
                Object {
                        "dispatch": [Function],
                        "getState": [Function],
                        "replaceReducer": [Function],
                        "subscribe": [Function],
                        Symbol(observable): [Function],
                      }
        }
        title="The form title"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Connect {
        "_reactInternalInstance": [Circular],
        "context": Object {
          "store": undefined,
        },
        "dispatchProps": Object {
          "closeModal": [Function],
          "openModal": [Function],
          "resetValues": [Function],
        },
        "doDispatchPropsDependOnOwnProps": false,
        "doStatePropsDependOnOwnProps": true,
        "finalMapDispatchToProps": [Function],
        "finalMapStateToProps": [Function],
        "hasStoreStateChanged": false,
        "haveOwnPropsChanged": false,
        "haveStatePropsBeenPrecalculated": false,
        "mergedProps": Object {
          "closeModal": [Function],
          "fields": Array [
            Object {
              "component": "TextInput",
              "variable": "foo",
            },
          ],
          "handleSubmit": [Function],
          "initialValues": Object {},
          "name": "baz",
          "openModal": [Function],
          "resetValues": [Function],
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "title": "The form title",
        },
        "props": Object {
          "fields": Array [
            Object {
              "component": "TextInput",
              "variable": "foo",
            },
          ],
          "handleSubmit": [Function],
          "name": "baz",
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "title": "The form title",
        },
        "refs": Object {},
        "renderedElement": <NodeForm
          addAnother={false}
          closeModal={[Function]}
          fields={
                    Array [
                              Object {
                                "component": "TextInput",
                                "variable": "foo",
                              },
                            ]
          }
          handleSubmit={[Function]}
          initialValues={Object {}}
          name="baz"
          node={Object {}}
          openModal={[Function]}
          resetValues={[Function]}
          store={
                    Object {
                              "dispatch": [Function],
                              "getState": [Function],
                              "replaceReducer": [Function],
                              "subscribe": [Function],
                              Symbol(observable): [Function],
                            }
          }
          title="The form title"
/>,
        "state": Object {
          "storeState": Object {
            "modals": Array [
              Object {
                "name": "baz",
                "open": true,
              },
            ],
            "protocol": Object {
              "config": Object {
                "registry": Object {
                  "foo": Object {
                    "label": "Foo",
                    "type": "string",
                  },
                },
              },
            },
          },
        },
        "stateProps": Object {
          "initialValues": Object {},
        },
        "statePropsPrecalculationError": null,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "version": 20,
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <NodeForm
          addAnother={false}
          closeModal={[Function]}
          fields={
                    Array [
                              Object {
                                "component": "TextInput",
                                "variable": "foo",
                              },
                            ]
          }
          handleSubmit={[Function]}
          initialValues={Object {}}
          name="baz"
          node={Object {}}
          openModal={[Function]}
          resetValues={[Function]}
          store={
                    Object {
                              "dispatch": [Function],
                              "getState": [Function],
                              "replaceReducer": [Function],
                              "subscribe": [Function],
                              Symbol(observable): [Function],
                            }
          }
          title="The form title"
/>,
        "_debugID": 2,
        "_renderedOutput": <NodeForm
          addAnother={false}
          closeModal={[Function]}
          fields={
                    Array [
                              Object {
                                "component": "TextInput",
                                "variable": "foo",
                              },
                            ]
          }
          handleSubmit={[Function]}
          initialValues={Object {}}
          name="baz"
          node={Object {}}
          openModal={[Function]}
          resetValues={[Function]}
          store={
                    Object {
                              "dispatch": [Function],
                              "getState": [Function],
                              "replaceReducer": [Function],
                              "subscribe": [Function],
                              Symbol(observable): [Function],
                            }
          }
          title="The form title"
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Connect(NodeForm)
    fields={
        Array [
            Object {
              "component": "TextInput",
              "variable": "foo",
            },
          ]
    }
    handleSubmit={[Function]}
    name="baz"
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
    title="The form title"
/>,
}
`;
